plugins/inputs/cloud_pubsub/pubsub.go

```diff
@@ -78,35 +84,37 @@ func (ps *PubSub) Start(ac telegraf.Accumulator) error {
 		return fmt.Errorf(`"project" is required`)
 	}
 
-	cctx, cancel := context.WithCancel(context.Background())
+	ps.sem = make(semaphore, ps.MaxUndeliveredMessages)
+	ps.acc = ac.WithTracking(ps.MaxUndeliveredMessages)
+
+	// Create top-level context with cancel that will be called on Stop().
+	ctx, cancel := context.WithCancel(context.Background())
+	ps.ctx = ctx
```

Comment:
Don't assign the context to the struct, instead pass it down to the goroutines below and save only the cancel function.